src/events.c

```diff
@@ -90,22 +90,51 @@ static struct events_node *new_events_node(struct event_t *event) {
 }
 
 void install_event(struct events_list *events, struct event_t *event) {
+	struct events_node *new_node = new_events_node(event);
+
 	if (events->head == NULL) {
 		// First event for this control
-		struct events_node *new_node = new_events_node(event);
 		events->head = new_node;
 		events->tail = new_node;
 		return;
 	}
 
-	// TODO: we need to remove existing events_node for the same event
-	// and support NULL event to only remove event.
+	if (strcmp(event->name, events->head->event->name) == 0) {
+		// clear_event(events->head->event);
+		struct events_node *new_head = events->head->next;
+		free(events->head);
+		events->head = new_head;
+
+		if (events->head == NULL) {
+			// The only event got removed
+			events->head = new_node;
+			events->tail = new_node;
+			return;
+		}
+	} else {
```

Comment:
Why this else? Even if the head event is not the same we are installing, the event is still to be installed...

Reply:
Pseudocode:
```c
if(list empty){
	head = tail = new event;
}

if(the head has the same name){
	remove head;
	if(list is empty){
		head = tail = new event;
	}
} else { // duplicate could still be in middle or tail
	for(event in events){
		if(event has the same name){
			remove event;
			if(was the last event){
				adjust tail;
			}
		}
	}
}

add new event to tail;
```

The first if case applies the head should be removed, the else case is for when a event should be removed in the middle or at the tail.

Reply:
ok, I got it!